src/resolution/composite-arns-resolver.ts

```diff
@@ -16,34 +16,70 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 import winston from 'winston';
-import { NameResolution, NameResolver } from '../types.js';
+import { KVBufferStore, NameResolution, NameResolver } from '../types.js';
+import * as metrics from '../metrics.js';
+import { KvArnsStore } from '../store/kv-arns-store.js';
 
 export class CompositeArNSResolver implements NameResolver {
   private log: winston.Logger;
   private resolvers: NameResolver[];
+  private cache: KVBufferStore;
 
   constructor({
     log,
     resolvers,
+    cache,
   }: {
     log: winston.Logger;
     resolvers: NameResolver[];
+    cache: KvArnsStore;
   }) {
-    this.log = log.child({ class: 'CompositeArNSResolver' });
+    this.log = log.child({ class: this.constructor.name });
     this.resolvers = resolvers;
+    this.cache = cache;
   }
 
   async resolve(name: string): Promise<NameResolution> {
     this.log.info('Resolving name...', { name });
 
     try {
+      const cachedResolutionBuffer = await this.cache.get(name);
+      if (cachedResolutionBuffer) {
+        const cachedResolution: NameResolution = JSON.parse(
+          cachedResolutionBuffer.toString(),
+        );
+        if (
+          cachedResolution !== undefined &&
+          cachedResolution.resolvedAt !== undefined &&
+          cachedResolution.ttl !== undefined &&
+          cachedResolution.resolvedAt + cachedResolution.ttl * 1000 > Date.now()
+        ) {
+          metrics.arnsCacheHitCounter.inc();
+          this.log.info('Cache hit for arns name', { name });
+          return cachedResolution;
+        }
+      }
+      metrics.arnsCacheMissCounter.inc();
+      this.log.info('Cache miss for arns name', { name });
+
       for (const resolver of this.resolvers) {
-        this.log.debug('Attempting to resolve name with resolver', {
-          resolver,
-        });
-        const resolution = await resolver.resolve(name);
-        if (resolution.resolvedId !== undefined) {
-          return resolution;
+        try {
+          this.log.info('Attempting to resolve name with resolver', {
+            type: resolver.constructor.name,
+            name,
+          });
+          const resolution = await resolver.resolve(name);
+          if (resolution.resolvedId !== undefined) {
+            await this.cache.set(name, Buffer.from(JSON.stringify(resolution)));
```

Comment:
I was thinking about the need to `await` the cache set here. For other caches we generally don't wait for it to be set.
In this case, I imagine if we call `resolve` for the same name several times (while the first is being cached) the sequential calls, while the name is not cached yet, will not wait for the cache, right? If so I don't think we should wait for the cache to be set.

What do you think?

Reply:
good call out - in addition to removing the `await` here, i realized we aren't handling concurrent requests well. I'd like to propose adding a request cache in the arns middleware that protects against concurrent calls to `naemResolver.resolve(name)` by storing the promise in a `NodeCache`. 

```typescript
    // check if this instance is already in the process of resolving the requested name, and return that promise if so, otherwise set it in the cache
    const getArnsResolutionPromise = async (): Promise<NameResolution> => {
      if (arnsRequestCache.has(arnsSubdomain)) {
        const arnsResolutionPromise =
          arnsRequestCache.get<Promise<NameResolution>>(arnsSubdomain);
        if (arnsResolutionPromise) {
          return arnsResolutionPromise;
        }
      }
      const arnsResolutionPromise = nameResolver.resolve(arnsSubdomain);
      arnsRequestCache.set(arnsSubdomain, arnsResolutionPromise);
      return arnsResolutionPromise;
    };

    const start = Date.now();
    try {
      // resolve the promise provided by the cache
      const { resolvedId, ttl, processId } = await getArnsResolutionPromise();
      metrics.arnsResolutionTime.observe(Date.now() - start);
      if (resolvedId === undefined) {
        sendNotFound(res);
        return;
      }
      res.header(headerNames.arnsResolvedId, resolvedId);
      res.header(headerNames.arnsTtlSeconds, ttl.toString());
      res.header(headerNames.arnsProcessId, processId);
      // TODO: add a header for arns cache status
      res.header('Cache-Control', `public, max-age=${ttl}`);
      dataHandler(req, res, next);
    } finally {
      // remove from cache after resolution
      arnsRequestCache.del(arnsSubdomain);
    }
```

with some added testing using `hey` - before the change

```bash
❯ hey -n 200 -c 200 -t 60 -host 'gateways.example.com' http://localhost:4000

Summary:
  Total:        20.2042 secs
  Slowest:      19.9149 secs
  Fastest:      0.0030 secs
  Average:      1.6396 secs
  Requests/sec: 9.8990
  
  Total data:   700772 bytes
  Size/request: 4273 bytes

Response time histogram:
  0.003 [1]     |
  1.994 [149]   |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  3.985 [0]     |
  5.977 [0]     |
  7.968 [0]     |
  9.959 [0]     |
  11.950 [0]    |
  13.941 [0]    |
  15.932 [0]    |
  17.924 [3]    |■
  19.915 [11]   |■■■


Latency distribution:
  10% in 0.0040 secs
  25% in 0.0174 secs
  50% in 0.0493 secs
  75% in 0.0605 secs
  90% in 0.1237 secs
  95% in 18.3069 secs
  99% in 19.9149 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0008 secs, 0.0030 secs, 19.9149 secs
  DNS-lookup:   0.0003 secs, 0.0000 secs, 0.0026 secs
  req write:    0.0001 secs, 0.0000 secs, 0.0015 secs
  resp wait:    1.6387 secs, 0.0030 secs, 19.9097 secs
  resp read:    0.0000 secs, 0.0000 secs, 0.0001 secs

Status code distribution:
  [200] 164 responses

Error distribution:
  [36]  Get "http://localhost:4000": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

And these logs are printed 200 times (one for each request)
```bash
 Resolving name... {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.996Z"}
core-1  | info: Cache miss for arns name {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.999Z"}
core-1  | info: Attempting to resolve name with resolver {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.999Z","type":"OnDemandArNSResolver"}
core-1  | info: Resolving name... {"class":"OnDemandArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:10.000Z"}
```

After the change:

❯ hey -n 200 -c 200 -t 60 -host 'gateways.example.com' http://localhost:4000

```bash
Summary:
  Total:        21.0792 secs
  Slowest:      20.3132 secs
  Fastest:      0.2127 secs
  Average:      5.2631 secs
  Requests/sec: 9.4880
  
  Total data:   1000 bytes
  Size/request: 5 bytes

Response time histogram:
  0.213 [1]     |
  2.223 [149]   |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  4.233 [0]     |
  6.243 [0]     |
  8.253 [0]     |
  10.263 [0]    |
  12.273 [0]    |
  14.283 [0]    |
  16.293 [0]    |
  18.303 [0]    |
  20.313 [50]   |■■■■■■■■■■■■■


Latency distribution:
  10% in 0.2267 secs
  25% in 0.2373 secs
  50% in 0.2655 secs
  75% in 20.2421 secs
  90% in 20.2759 secs
  95% in 20.2884 secs
  99% in 20.2986 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0014 secs, 0.2127 secs, 20.3132 secs
  DNS-lookup:   0.0006 secs, 0.0000 secs, 0.0028 secs
  req write:    0.0001 secs, 0.0000 secs, 0.0018 secs
  resp wait:    5.2616 secs, 0.2126 secs, 20.3059 secs
  resp read:    0.0000 secs, 0.0000 secs, 0.0001 secs

Status code distribution:
  [200] 200 responses
```

Logs (only printed once)
```bash
 Resolving name... {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.996Z"}
core-1  | info: Cache miss for arns name {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.999Z"}
core-1  | info: Attempting to resolve name with resolver {"class":"CompositeArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:09.999Z","type":"OnDemandArNSResolver"}
core-1  | info: Resolving name... {"class":"OnDemandArNSResolver","name":"gateways","timestamp":"2024-09-10T13:34:10.000Z"}
```

This should also help slamming AO on fresh or expired names with only one resolution request and likely a pattern would could extend to other request paths.

cc @djwhitt